### ISSUE TITLE
Invalid left hand side messages are unhelpful

### ISSUE THREAD

**Issue by JsonFreeman** [microsoft/TypeScript#2474]
``` ts
class C { }
enum A { X }
C = 0; // Invalid left-hand side of assignment expression.
A.X = 0; // Invalid left-hand side of assignment expression.
A.X++; // The operand of an increment or decrement operator must be a variable, property or indexer.
```

We give different error messages for different syntactic constructs that can mutate something (assignment, increment, for...in, etc). But the syntactic use of the entity is completely obvious. Yet, we don't give different error messages for different kinds of entities that are illegal to mutate (enum member, class, import, etc). For the user, the reason the thing can't be mutated may still be a mystery, and the user needs to know what kind of entity they were trying to mutate.

The message does not have to vary based on the syntactic mutation construct. But it should vary based on the entity that cannot be mutated.

**Comment by mhegazy** [comment]
What is the suggested error message(s)?

**Comment by vladima** [comment]
IMO error message should explain why mutation is not legal i.e. 'Enum member is not a valid left-hand side of an assignment' or ''arguments is not a valid operand of increment\decrement operator', whereis now we just  telling that we are going to mutate something and it is not allowed but don't explain why. This sometimes can be confusing i.e.

```ts
A.B++ // errors: The operand of an increment or decrement operator must be a variable, property or indexer. 
```

But syntactically is does look like a property access so I have to go to the declaration of `A` to figure out that this is enum. It will be nicer if error message will give this information right of the bat.

**Comment by JsonFreeman** [comment]
Something like "Cannot change the value of an enum member", or "cannot change the value of an import"

**Comment by shramee** [comment]
Yeah, would b really nice to have a more explicit one.

**Comment by JoshuaKGoldberg** [comment]
Looks like this issue has since been resolved, mostly? The current error messages stop just short of delineating enum members from class properties:

```
class C { }
enum A { X }
C = 0; // Cannot assign to 'C' because it is not a variable.
A.X = 0; // Cannot assign to 'X' because it is a read-only property.
A.X++; // Cannot assign to 'X' because it is a read-only property.
```
